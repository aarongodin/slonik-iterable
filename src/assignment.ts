import { sql, SqlTokenType } from "slonik"

import { translateValue } from "./translate"

export function fromObject(payload: Record<string, any>, translate = translateValue): SqlTokenType {
  const values = Object.values(
    Object.entries(payload).map(([column, value]) => {
      return sql`${sql.identifier([column])} = ${translate(column, value)}`
    }),
  )
  return sql.join(values, sql`, `)
}
