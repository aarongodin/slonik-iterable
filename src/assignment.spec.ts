import { sql } from "slonik"

import * as assignment from "./assignment"

describe("assignment helpers", () => {
  describe("fromObject()", () => {
    it("should return a constructed SqlTokenType", () => {
      const actual = assignment.fromObject({
        first: "my first value",
      })

      expect(actual).toMatchSnapshot()
    })

    it("should return a constructed SqlTokenType with basic type assumptions", () => {
      const actual = assignment.fromObject({
        first: "a string",
        second: 22,
      })

      expect(actual).toMatchSnapshot()
    })

    describe("given a custom translate function", () => {
      it("should return a constructed SqlTokenType", () => {
        const actual = assignment.fromObject(
          {
            customType: { iamJson: true },
            customType2: ["one", "two", "three"],
          },
          (column, value) => {
            if (column === "customType") {
              return sql.json(value)
            } else if (column === "customType2") {
              return sql.array(value, "text")
            } else {
              return value
            }
          },
        )

        expect(actual).toMatchSnapshot()
      })
    })
  })
})
