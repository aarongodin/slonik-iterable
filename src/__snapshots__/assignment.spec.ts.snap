// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignment helpers fromObject() given a custom translate function should return a constructed SqlTokenType 1`] = `
Object {
  "glue": Object {
    "sql": ", ",
    "type": "SLONIK_TOKEN_SQL",
    "values": Array [],
  },
  "members": Array [
    Object {
      "sql": "\\"customType\\" = $1",
      "type": "SLONIK_TOKEN_SQL",
      "values": Array [
        "{\\"iamJson\\":true}",
      ],
    },
    Object {
      "sql": "\\"customType2\\" = $1::\\"text\\"[]",
      "type": "SLONIK_TOKEN_SQL",
      "values": Array [
        Array [
          "one",
          "two",
          "three",
        ],
      ],
    },
  ],
  "type": "SLONIK_TOKEN_LIST",
}
`;

exports[`assignment helpers fromObject() should return a constructed SqlTokenType 1`] = `
Object {
  "glue": Object {
    "sql": ", ",
    "type": "SLONIK_TOKEN_SQL",
    "values": Array [],
  },
  "members": Array [
    Object {
      "sql": "\\"first\\" = $1",
      "type": "SLONIK_TOKEN_SQL",
      "values": Array [
        "my first value",
      ],
    },
  ],
  "type": "SLONIK_TOKEN_LIST",
}
`;

exports[`assignment helpers fromObject() should return a constructed SqlTokenType with basic type assumptions 1`] = `
Object {
  "glue": Object {
    "sql": ", ",
    "type": "SLONIK_TOKEN_SQL",
    "values": Array [],
  },
  "members": Array [
    Object {
      "sql": "\\"first\\" = $1",
      "type": "SLONIK_TOKEN_SQL",
      "values": Array [
        "a string",
      ],
    },
    Object {
      "sql": "\\"second\\" = $1",
      "type": "SLONIK_TOKEN_SQL",
      "values": Array [
        22,
      ],
    },
  ],
  "type": "SLONIK_TOKEN_LIST",
}
`;
